src/openms/source/FILTERING/DATAREDUCTION/Deisotoper.cpp

```diff
@@ -38,9 +38,341 @@
 #include <OpenMS/KERNEL/MSSpectrum.h>
 #include <OpenMS/MATH/MISC/MathFunctions.h>
+#include <OpenMS/CHEMISTRY/ISOTOPEDISTRIBUTION/CoarseIsotopePatternGenerator.h>
+#include <OpenMS/FILTERING/TRANSFORMERS/ThresholdMower.h>
+#include <OpenMS/FILTERING/TRANSFORMERS/NLargest.h>
 
 namespace OpenMS
 {
+// static
+void Deisotoper::deisotopeWithAveragineModel(MSSpectrum& spec,
+  double fragment_tolerance,
+  bool fragment_unit_ppm,
+  bool rem_low_intensity,
+  int min_charge,
+  int max_charge,
+  bool keep_only_deisotoped,
+  unsigned int min_isopeaks,
+  unsigned int max_isopeaks,
+  bool make_single_charged,
+  bool annotate_charge,
+  bool annotate_iso_peak_count,
+  bool add_up_intensity,
+  bool used_for_open_search)
+{
+  OPENMS_PRECONDITION(spec.isSorted(), "Spectrum must be sorted.");
+
+  if (min_isopeaks < 2 || max_isopeaks < 2 || min_isopeaks > max_isopeaks)
+  {
+    throw Exception::IllegalArgument(__FILE__,
+      __LINE__,
+      OPENMS_PRETTY_FUNCTION,
+      "Minimum/maximum number of isotopic peaks must be at least 2 (and min_isopeaks <= max_isopeaks).");
+  }
+
+  if (spec.empty()) { return; }
+
+  // remove 0 intensity peaks
+  ThresholdMower threshold_mower_filter;
+  threshold_mower_filter.filterPeakSpectrum(spec);
+
+  // discard low-intensity peaks
+  if (rem_low_intensity)
+  { 
+    UInt max_num_peaks = used_for_open_search ? 1000 : 5000;
+
+    // only keep max_num_peaks highest peaks
+    NLargest nlargest_filter = NLargest(max_num_peaks);
+    nlargest_filter.filterPeakSpectrum(spec);
+    
+    spec.sortByPosition();
+  }
+
+  // store additional information if specified
+  Size charge_index(0);
+  Size iso_peak_count_index(0);
+
+  // reserve integer data array to store charge of peaks
+  if (annotate_charge)
+  {
+    // expand to hold one additional integer data array to hold the charge
+    spec.getIntegerDataArrays().resize(spec.getIntegerDataArrays().size() + 1);
+    spec.getIntegerDataArrays().back().setName("charge");
+    charge_index = spec.getIntegerDataArrays().size() - 1;
+  }
+  // reserve integer data array to store number of isotopic peaks for each isotopic pattern
+  if (annotate_iso_peak_count)
+  {
+    spec.getIntegerDataArrays().resize(spec.getIntegerDataArrays().size() + 1);
+    spec.getIntegerDataArrays().back().setName("iso_peak_count");
+    iso_peak_count_index = spec.getIntegerDataArrays().size() - 1;
+  }
+
+  // during discovery phase, work on a constant reference (just to make sure we do not modify spec)
+  const MSSpectrum& old_spectrum = spec;
+
+  // determine charge seeds and extend them
+  std::vector<size_t> mono_isotopic_peak_charge(old_spectrum.size(), 0);
+  std::vector<int> features(old_spectrum.size(), -1);
+  std::vector<double> mono_iso_peak_intensity(old_spectrum.size(), 0);
+  std::vector<Size> iso_peak_count(old_spectrum.size(), 1);
+  int feature_number = 0;
+
+  std::vector<size_t> extensions;
+  std::vector< std::vector<size_t> > clusters;
```

Comment:
this will be extremely inefficient in terms of runtime. If you have a maximum number of peaks inside the inner vector (max_isopeaks?) then flatten to vector<size_t> and manage the number of clusters using a counter.
May not dominate runtime though...